[BUG] Switching nozzle tool change mouvement error

Using a switching nozzle and extruder using the same servo ( https://www.thingiverse.com/thing:673816 ) when executing a tool change, the tool's mouvement are not the ones expected:

Expected moves:
-Raise z by TOOLCHANGE_ZRAISE 
-Rotate servo
-Move head to compensate HOTEND_OFFSET_X ,Y and Z
-Lower z by TOOLCHANGE_ZRAISE 

Actual moves:
-Rotate servo
-Lower z by a certain distance, which isn't TOOLCHANGE_ZRAISE

I am using the latest bugfix 2.0.x 

[Configuration.zip](https://github.com/MarlinFirmware/Marlin/files/4017240/Configuration.zip)
[Configuration_adv.zip](https://github.com/MarlinFirmware/Marlin/files/4017241/Configuration_adv.zip)

Compilation error. :-(
I compiled without the updates of libraries and platform (stm32 5.7.0)

Marlin\src\feature\pause.cpp: In function 'bool unload_filament(const float&, bool, PauseMode)':
Marlin\src\feature\pause.cpp:355:20: error: 'FILAMENT_UNLOAD_PURGE_FEEDRATE' was not declared in this scope
                   (FILAMENT_UNLOAD_PURGE_FEEDRATE) * mix_multiplier);
                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Marlin\src\feature\pause.cpp:355:20: note: suggested alternative: 'FILAMENT_UNLOAD_PURGE_RETRACT'
                   (FILAMENT_UNLOAD_PURGE_FEEDRATE) * mix_multiplier);
                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                    FILAMENT_UNLOAD_PURGE_RETRACT
*** [.pio\build\BIGTREE_SKR_PRO\src\src\feature\pause.cpp.o] Error 1
=================== [FAILED] Took 32.62 seconds 
Environment                   Status    Duration
----------------------------  --------  ------------
megaatmega2560                IGNORED
...
STM32F407VE_black             IGNORED
BIGTREE_SKR_PRO               FAILED    00:00:32.620
BIGTREE_BTT002                IGNORED
...
=== 1 failed, 0 succeeded in 00:00:32.620

--- a/Marlin/src/module/tool_change.cpp
+++ b/Marlin/src/module/tool_change.cpp
@@ -930,20 +930,17 @@ void tool_change(const uint8_t new_tool, bool no_move/*=false*/) {
         // Raise by a configured distance to avoid workpiece, except with
         // SWITCHING_NOZZLE_TWO_SERVOS, as both nozzles will lift instead.
         if (!no_move) {
+          #if HAS_SOFTWARE_ENDSTOPS
+            const float maxz = _MIN(soft_endstop.max.z, Z_MAX_POS);
+          #else
+            constexpr float maxz = Z_MAX_POS;
+          #endif
+
+          // Check if Z has space to compensate at least z_offset, and if not, just abort now
           const float newz = current_position.z + _MAX(-diff.z, 0.0);
+          if (newz > maxz) return;
 
-          #if HAS_SOFTWARE_ENDSTOPS
-            // Check if Z has space to compensate at least z_offset, and if not, just abort now
-            if (newz > soft_endstop.max.z) return;
-          #endif
-
-          newz += toolchange_settings.z_raise;
-
-          #if HAS_SOFTWARE_ENDSTOPS
-            NOMORE(newz, soft_endstop.max.z);
-          #endif
-
-          current_position.z = newz;
+          current_position.z = _MIN(newz + toolchange_settings.z_raise, maxz);
           fast_line_to_current(Z_AXIS);
         }
         move_nozzle_servo(new_tool);

--- a/Marlin/src/module/tool_change.cpp
+++ b/Marlin/src/module/tool_change.cpp
@@ -931,6 +931,9 @@ void tool_change(const uint8_t new_tool, bool no_move/*=false*/) {
         // SWITCHING_NOZZLE_TWO_SERVOS, as both nozzles will lift instead.
         current_position.z += _MAX(-diff.z, 0.0);
         if (!no_move) {
+          // Check if Z has space to compensate at least z_offset, if not enough space abort tool change
+          if (current_position.z > soft_endstop.max.z) return
+
           current_position.z += toolchange_settings.z_raise;
           #if HAS_SOFTWARE_ENDSTOPS
             NOMORE(current_position.z, soft_endstop.max.z);

--- a/Marlin/src/module/tool_change.cpp
+++ b/Marlin/src/module/tool_change.cpp
@@ -931,8 +931,10 @@ void tool_change(const uint8_t new_tool, bool no_move/*=false*/) {
         // SWITCHING_NOZZLE_TWO_SERVOS, as both nozzles will lift instead.
         current_position.z += _MAX(-diff.z, 0.0);
         if (!no_move) {
-          // Check if Z has space to compensate at least z_offset, if not enough space abort tool change
-          if (current_position.z > soft_endstop.max.z) return
+          #if HAS_SOFTWARE_ENDSTOPS
+            // Check if Z has space to compensate at least z_offset, if not enough space abort tool change
+            if (current_position.z > soft_endstop.max.z) return
+          #endif
 
           current_position.z += toolchange_settings.z_raise;
           #if HAS_SOFTWARE_ENDSTOPS

--- a/Marlin/src/module/tool_change.cpp
+++ b/Marlin/src/module/tool_change.cpp
@@ -929,11 +929,14 @@ void tool_change(const uint8_t new_tool, bool no_move/*=false*/) {
       #elif ENABLED(SWITCHING_NOZZLE) && !SWITCHING_NOZZLE_TWO_SERVOS   // Switching Nozzle (single servo)
         // Raise by a configured distance to avoid workpiece, except with
         // SWITCHING_NOZZLE_TWO_SERVOS, as both nozzles will lift instead.
-        current_position.z += _MAX(-diff.z, 0.0) + toolchange_settings.z_raise;
-        #if HAS_SOFTWARE_ENDSTOPS
-          NOMORE(current_position.z, soft_endstop.max.z);
-        #endif
-        if (!no_move) fast_line_to_current(Z_AXIS);
+        current_position.z += _MAX(-diff.z, 0.0);
+        if (!no_move) {
+          current_position.z += toolchange_settings.z_raise;
+          #if HAS_SOFTWARE_ENDSTOPS
+            NOMORE(current_position.z, soft_endstop.max.z);
+          #endif
+          fast_line_to_current(Z_AXIS);
+        }
         move_nozzle_servo(new_tool);
       #endif
 

--- a/Marlin/src/module/tool_change.cpp
+++ b/Marlin/src/module/tool_change.cpp
@@ -929,8 +929,8 @@ void tool_change(const uint8_t new_tool, bool no_move/*=false*/) {
       #elif ENABLED(SWITCHING_NOZZLE) && !SWITCHING_NOZZLE_TWO_SERVOS   // Switching Nozzle (single servo)
         // Raise by a configured distance to avoid workpiece, except with
         // SWITCHING_NOZZLE_TWO_SERVOS, as both nozzles will lift instead.
-        const float newz = current_position.z + _MAX(-diff.z, 0.0);
         if (!no_move) {
+          const float newz = current_position.z + _MAX(-diff.z, 0.0);
 
           #if HAS_SOFTWARE_ENDSTOPS
             // Check if Z has space to compensate at least z_offset, and if not, just abort now

--- a/Marlin/src/module/tool_change.cpp
+++ b/Marlin/src/module/tool_change.cpp
@@ -929,17 +929,21 @@ void tool_change(const uint8_t new_tool, bool no_move/*=false*/) {
       #elif ENABLED(SWITCHING_NOZZLE) && !SWITCHING_NOZZLE_TWO_SERVOS   // Switching Nozzle (single servo)
         // Raise by a configured distance to avoid workpiece, except with
         // SWITCHING_NOZZLE_TWO_SERVOS, as both nozzles will lift instead.
-        current_position.z += _MAX(-diff.z, 0.0);
+        const float newz = current_position.z + _MAX(-diff.z, 0.0);
         if (!no_move) {
+
           #if HAS_SOFTWARE_ENDSTOPS
-            // Check if Z has space to compensate at least z_offset, if not enough space abort tool change
-            if (current_position.z > soft_endstop.max.z) return
+            // Check if Z has space to compensate at least z_offset, and if not, just abort now
+            if (newz > soft_endstop.max.z) return;
           #endif
 
-          current_position.z += toolchange_settings.z_raise;
+          newz += toolchange_settings.z_raise;
+
           #if HAS_SOFTWARE_ENDSTOPS
-            NOMORE(current_position.z, soft_endstop.max.z);
+            NOMORE(newz, soft_endstop.max.z);
           #endif
+
+          current_position.z = newz;
           fast_line_to_current(Z_AXIS);
         }
         move_nozzle_servo(new_tool);
@@ -950,7 +954,7 @@ void tool_change(const uint8_t new_tool, bool no_move/*=false*/) {
       #endif
 
       // The newly-selected extruder XYZ is actually at...
-      if (DEBUGGING(LEVELING)) DEBUG_ECHOLNPAIR("Offset Tool XY by { ", diff.x, ", ", diff.y, ", ", diff.z, " }");
+      if (DEBUGGING(LEVELING)) DEBUG_ECHOLNPAIR("Offset Tool XYZ by { ", diff.x, ", ", diff.y, ", ", diff.z, " }");
       current_position += diff;
 
       // Tell the planner the new "current position"

